pkg/volume/rbd/rbd_util.go

```
@@ -0,0 +1,145 @@
+	return path.Join(host.GetPluginDir(RBDPluginName), "rbd", pool+"-image-"+image)
+}
+
+type RBDUtil struct{}
+
+func (util *RBDUtil) MakeGlobalPDName(rbd rbd) string {
+	return makePDNameInternal(rbd.plugin.host, rbd.pool, rbd.image)
+}
+
+func (util *RBDUtil) AttachDisk(rbd rbd) error {
+	var err error
+	devicePath := strings.Join([]string{"/dev/rbd", rbd.pool, rbd.image}, "/")
+	exist := waitForPathToExist(devicePath, 1)
+	if exist == false {
+		// modprobe
+		_, err = rbd.plugin.execCommand("modprobe", []string{"rbd"})
+		if err != nil {
+			glog.Errorf("rbd: failed to modprobe rbd error:%v", err)
+			return err
+		}
+		// rbd map
+		l := len(rbd.mon)
+		// avoid mount storm, pick a host randomly
+		start := rand.Int() % l
+		// iterate all hosts until mount succeeds.
+		for i := start; i < start+l; i++ {
+			mon := rbd.mon[i%l]
+			glog.V(1).Infof("rbd: map mon %s", mon)
+			if rbd.secret != "" {
+				_, err = rbd.plugin.execCommand("rbd",
+					[]string{"map", rbd.image, "--pool", rbd.pool, "--id", rbd.id, "-m", mon, "--key=" + rbd.secret})
+			} else {
+				_, err = rbd.plugin.execCommand("rbd",
+					[]string{"map", rbd.image, "--pool", rbd.pool, "--id", rbd.id, "-m", mon, "-k", rbd.keyring})
+			}
+			if err == nil {
+				break
+			}
+		}
+	}
+	if err != nil {
+		glog.Errorf("rbd: failed to map image:Error: %v", err)
+		return err
+	}
+	exist = waitForPathToExist(devicePath, 10)
+	if !exist {
+		return errors.New("Could not map image: Timeout after 10s")
+	}
+	// mount it
+	globalPDPath := rbd.manager.MakeGlobalPDName(rbd)
+	mountpoint, err := rbd.mounter.IsMountPoint(globalPDPath)
+	if mountpoint {
+		glog.Infof("rbd: %s already mounted", globalPDPath)
+		return nil
+	}
+
+	if err := os.MkdirAll(globalPDPath, 0750); err != nil {
+		glog.Errorf("rbd: failed to mkdir %s, error", globalPDPath)
+		return err
+	}
+
+	err = rbd.mounter.Mount(devicePath, globalPDPath, rbd.fsType, nil)
```

This should probably be a single block error (if err := ...) and in the block you should transform or return the error instead of logging (if you want the info to be included, wrap the error, otherwise return straight for this function.